internal/server/statvar/statvar_hierarchy_util.go

```diff
@@ -164,45 +171,165 @@ func BuildStatVarSearchIndex(
 // A BleveDocument models a document by the bleve index.
 // Currently we index stat vars and treat them as documents.
 type BleveDocument struct {
+	// Id of the statvar.
+	Id string `index:"false"`
 	// Title of the document. For a statvar this will be the DisplayName.
-	Title string
+	Title     string `json:"t"`
+	KeyText   string `json:"k" store:"false"`
+	ValueText string `json:"v" store:"false"`
 	// A key value pairs string describing the properties of a stat var.
-	KeyValueText string
+	KeyValueText string `json:"kv" store:"false"`
+	// Searchable name of the statvar.
+	SearchName string `json:"sn" store:"false"`
+	// Number of constraints associated to a statvar.
+	NumConstraints int32 `json:"nc" store:"false"`
+	// Number of title tokens, the fewer the more generic a statvar.
+	NumTitleTokens int32 `json:"nt" store:"false"`
 }
 
-// BuildBleveIndex builds the bleve search index for all the stat vars.
-func BuildBleveIndex(
-	rawSvg map[string]*pb.StatVarGroupNode,
-) (bleve.Index, error) {
-	defer util.TimeTrack(time.Now(), "BuildBleveIndex")
-	indexMapping := bleve.NewIndexMapping()
-
+func buildBleveDocumentMapping() *mapping.DocumentMapping {
 	documentMapping := bleve.NewDocumentMapping()
-	indexMapping.AddDocumentMapping("Document", documentMapping)
+	documentMapping.Dynamic = false
 
-	titleFieldMapping := bleve.NewTextFieldMapping()
-	titleFieldMapping.Store = true
-	documentMapping.AddFieldMappingsAt("Title", titleFieldMapping)
+	emptyDocument := BleveDocument{}
+	documentType := reflect.TypeOf(emptyDocument)
+	for i := 0; i < documentType.NumField(); i++ {
+		field := documentType.Field(i)
+		if field.Tag.Get("index") == "false" {
+			continue
+		}
+		var fieldMapping mapping.FieldMapping
 
-	keyValueTextFieldMapping := bleve.NewTextFieldMapping()
-	keyValueTextFieldMapping.Store = false
-	documentMapping.AddFieldMappingsAt("KeyValueText", keyValueTextFieldMapping)
+		if field.Type.Name() == "string" {
+			fieldMapping = *bleve.NewNumericFieldMapping()
+		} else {
+			fieldMapping = *bleve.NewTextFieldMapping()
+		}
+		if field.Tag.Get("store") == "false" {
+			fieldMapping.Store = false
+		}
+		documentMapping.AddFieldMappingsAt(field.Name, &fieldMapping)
+	}
+	return documentMapping
+}
 
-	index, err := bleve.NewUsing("", indexMapping, bleve.Config.DefaultIndexType, bleve.Config.DefaultMemKVStore, nil)
+func buildBleveIndexMapping() (mapping.IndexMapping, error) {
+	indexMapping := bleve.NewIndexMapping()
+
+	// Lowercase all the tokens and use a custom list of stopwords.
+	err := indexMapping.AddCustomTokenFilter("lowercase", map[string]interface{}{
+		"type": lowercase.Name,
+	})
 	if err != nil {
 		return nil, err
 	}
-	batch := index.NewBatch()
+	err = indexMapping.AddCustomTokenMap("tokenmap_custom_stopwords", map[string]interface{}{
+		"type": tokenmap.Name,
+		"tokens": []interface{}{
```

Comment:
Move the list of tokens here to the top of file as a constant

Reply:
Apparently you can't define constant arrays in golang.